File: Sortings/InsertionSort/InsertionSort.java

public class InsertionSort {

    /**
     * INSERTION SORT
     *
     * Idea: - Treat the left portion of the array as sorted. - Take nums[i],
     * insert it into its correct position in the sorted left side.
     *
     * Time Complexity: - Best case: O(n) (already sorted) - Worst/Avg: O(n^2)
     *
     * Space Complexity: O(1)
     *
     * Stable: Yes
     *
     * Works well for: - Nearly sorted arrays (very fast) - Small arrays
     */
    public static void insertionSort(int[] nums) {
        int n = nums.length;

        for (int i = 1; i < n; i++) {
            int key = nums[i];
            int j = i - 1;

            // Shift elements larger than key to the right
            while (j >= 0 && nums[j] > key) {
                nums[j + 1] = nums[j];
                j--;
            }

            // Place key in correct position
            nums[j + 1] = key;
        }
    }

    public static void main(String[] args) {
        int[] arr = {12, 11, 13, 5, 6};

        insertionSort(arr);

        for (int x : arr) {
            System.out.print(x + " ");
        }
    }
}

File: Sortings/InsertionSort/InsertionSort.ts
/**
 * INSERTION SORT
 *
 * Idea:
 * - Assume left side is sorted.
 * - Insert arr[i] into correct sorted position in the left side.
 *
 * Time:
 * - Best: O(n)
 * - Worst: O(n^2)
 *
 * Space: O(1)
 * Stable: Yes
 */
export function insertionSort(arr: number[]): number[] {
  const nums = [...arr];

  for (let i = 1; i < nums.length; i++) {
    let key = nums[i];
    let j = i - 1;

    // Shift all bigger elements to the right
    while (j >= 0 && nums[j] > key) {
      nums[j + 1] = nums[j];
      j--;
    }

    nums[j + 1] = key;
  }

  return nums;
}

/* TEST */
console.log(insertionSort([12, 11, 13, 5, 6]));
